File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int left = 0, right = n - 1;
        int largest_from_left = 0, largest_from_right = 0;
        int trapped_water = 0;

        while (left < right) {
            largest_from_left = max(largest_from_left, height[left]);
            largest_from_right = max(largest_from_right, height[right]);

            if (largest_from_left <= largest_from_right) {
                left++;
                trapped_water += max(0, largest_from_left - height[left]);
            } else {
                right--;
                trapped_water += max(0, largest_from_right - height[right]);
            }
        }
        return trapped_water;
    }
};

File: Data Structures & Algorithms/house-robber-iii/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        # Should store each node and the max profit from robbing and max profix if don't rob
        dp = {}

        def dfs(node):
            if not node:
                return 0
            
            if node in dp:
                return dp[node]
            
            # Max profit from robbing
            rob = node.val
            if node.left:
                rob += dfs(node.left.left) + dfs(node.left.right)
            if node.right:
                rob += dfs(node.right.left) + dfs(node.right.right)

            dont_rob = dfs(node.left) + dfs(node.right)

            dp[node] = max(rob, dont_rob)

            return dp[node]
        
        return dfs(root)